### ISSUE TITLE
passing anonymous struct literal with particular nesting *after* comptime arg generates SIGBUS/misaligned read

### ISSUE THREAD

**Issue by kivikakk** [ziglang/zig#7534]
Might be the same as #5498. Here's the repro; I've condensed it as much as I can but it might be reduced further yet:

```zig
const U = union(enum) {
    A: u32,
    S: *const S,
};

const S = struct {
    u: U,
};

fn inspect(comptime a: u8, b: S) void {
    @import("std").debug.print("hi\n", .{});
}

pub fn main() !void {
    const sub_const = S{ .u = U{ .A = 0xcccccccc } };
    inspect(0, S{ .u = U{ .S = &sub_const } });
    inspect(0, .{ .u = U{ .S = &sub_const } });

    var sub_var = sub_const;
    inspect(0, S{ .u = U{ .S = &sub_var } });
    inspect(0, .{ .u = U{ .S = &sub_var } });
}
```

The 4th `inspect` call crashes with SIGBUS (Misaligned address error).

Moving `comptime a: u8` to be **after** `b: S` in the parameter list stops the crash. Removing the struct `S` nested between union layers and using `S: *const U` instead also stops the crash.

Changing `comptime a: u8` to `comptime a: void` doesn't stop the crash. Any comptime argument seems to do the trick.

Output on amd64 and arm64:

```console
$ set zig ~/Code/zig/build/bin/zig; file $zig; $zig run example.zig
/Users/kameliya/Code/zig/build/bin/zig: Mach-O 64-bit executable x86_64
hi
hi
hi
fish: Job 2, '$zig run example.zig' terminated by signal SIGBUS (Misaligned address error)
$ set zig ~/Code/zig/build-native/bin/zig; file $zig; $zig run example.zig
/Users/kameliya/Code/zig/build-native/bin/zig: Mach-O 64-bit executable arm64
hi
hi
hi
fish: Job 2, '$zig run example.zig' terminated by signal SIGBUS (Misaligned address error)
$ 
```

I've included lldb sessions of both crashes here: https://gist.github.com/kivikakk/3aaca18c136f6681146e777451fe4014

In both cases lldb nicely highlights the `.` in `U{ .S =` as the problem:

<img width="573" alt="image" src="https://user-images.githubusercontent.com/1915/103048729-a7339380-45e3-11eb-98ee-6fbef172e21b.png">

LLVM IR for `main`, first as written above, second if we remove the `comptime a: u8` parameter entirely: https://gist.github.com/kivikakk/61f642785dc6b5b94c9f894d7195f77e

**Comment by kivikakk** [comment]
Comparing the LLVM IR, with the `comptime` argument the call is:

```llvm
  store i1 true, i1* getelementptr inbounds (%"struct:21:17", %"struct:21:17"* @63, i32 0, i32 0, i32 1), align 1, !dbg !1638
  store %S* %sub_var, %S** getelementptr inbounds (%"struct:21:17", %"struct:21:17"* @63, i32 0, i32 0, i32 0), align 8, !dbg !1639
  call fastcc void @inspect(%S* @64), !dbg !1640
```

And without:

```llvm
  %8 = getelementptr inbounds %S, %S* %3, i32 0, i32 0, !dbg !1638
  %9 = getelementptr inbounds %U, %U* %8, i32 0, i32 1, !dbg !1639
  store i1 true, i1* %9, align 1, !dbg !1639
  %10 = getelementptr inbounds %U, %U* %8, i32 0, i32 0, !dbg !1639
  store %S* %sub_var, %S** %10, align 8, !dbg !1640
  call fastcc void @inspect(%S* %3), !dbg !1641
```

I don't really know LLVM IR, but it looks like the with-`comptime`-argument variant tries to call the function with `@64` after writing values into `@63`, which are constants:

```llvm
@63 = internal unnamed_addr constant %"struct:21:17" { %U { %S* undef, i1 true } }, align 8
@64 = internal unnamed_addr constant %S { %U { %S* undef, i1 true } }, align 8
```

**Comment by kivikakk** [comment]
I thought I found a new bug today and realised it was probably just this again:

```zig
const std = @import("std");
const testing = std.testing;

pub const Prop = union(enum) {
    Indirect: u8,
    Direct: u8,

    pub fn format(prop: Prop, comptime fmt: []const u8, options: std.fmt.FormatOptions, writer: anytype) !void {
        switch (prop) {
            .Indirect => |v| {
                const formatter = Struct{ .val = v };
                try std.fmt.format(writer, "indirect: {}", .{formatter});
            },
            .Direct => |v| try std.fmt.format(writer, "direct: {}", .{Struct{ .val = v }}),
        }
    }

    const Struct = struct {
        val: u8,
    };
};

test "succeeds" {
    var compatible = Prop{ .Indirect = 8 };
    std.debug.print("{}\n", .{compatible});
}

test "crashes" {
    var clock_names = Prop{ .Direct = 8 };
    std.debug.print("{}\n", .{clock_names});
}

test "also crashes" {
    std.debug.print("{}\n", .{Prop{ .Indirect = 8 }});
}
```

**Comment by mikdusan** [comment]
with latest compiler:
- repro no longer crashes and executes cleanly
- repro2 test does not crash
- it appears { repro, repro2} were fixed between 0.9.1 and 0.10.0

<details>
<summary>repro.zig</summary>

```zig
const U = union(enum) {
    A: u32,
    S: *const S,
};

const S = struct {
    u: U,
};

fn inspect(comptime _: u8, _: S) void {
    @import("std").debug.print("hi\n", .{});
}

pub fn main() !void {
    const sub_const = S{ .u = U{ .A = 0xcccccccc } };
    inspect(0, S{ .u = U{ .S = &sub_const } });
    inspect(0, .{ .u = U{ .S = &sub_const } });

    var sub_var = sub_const;
    inspect(0, S{ .u = U{ .S = &sub_var } });
    inspect(0, .{ .u = U{ .S = &sub_var } });
}
```
</details>

<details>
<summary>repro2.zig</summary>

```zig
const std = @import("std");
const testing = std.testing;

pub const Prop = union(enum) {
    Indirect: u8,
    Direct: u8,

    pub fn format(prop: Prop, comptime _: []const u8, _: std.fmt.FormatOptions, writer: anytype) !void {
        switch (prop) {
            .Indirect => |v| {
                const formatter = Struct{ .val = v };
                try std.fmt.format(writer, "indirect: {}", .{formatter});
            },
            .Direct => |v| try std.fmt.format(writer, "direct: {}", .{Struct{ .val = v }}),
        }
    }

    const Struct = struct {
        val: u8,
    };
};

test "succeeds" {
    var compatible = Prop{ .Indirect = 8 };
    std.debug.print("{}\n", .{compatible});
}

test "crashes" {
    var clock_names = Prop{ .Direct = 8 };
    std.debug.print("{}\n", .{clock_names});
}

test "also crashes" {
    std.debug.print("{}\n", .{Prop{ .Indirect = 8 }});
}
```
</details>

```sh
$ uname -a
Linux archlinux 6.1.2-arch1-1 #1 SMP PREEMPT_DYNAMIC Sat, 31 Dec 2022 17:40:35 +0000 x86_64 GNU/Linux

$ zig version
0.11.0-dev.1241+0507ced8c

$ zig run repro.zig
hi
hi
hi
hi

$ zig test repro2.zig
Test [1/3] test.succeeds... indirect: repro2.Prop.Struct{ .val = 8 }
Test [2/3] test.crashes... direct: repro2.Prop.Struct{ .val = 8 }
Test [3/3] test.also crashes... indirect: repro2.Prop.Struct{ .val = 8 }
All 3 tests passed.
```
